core/src/main/java/org/elasticsearch/index/shard/RefreshListeners.java

```diff
@@ -95,12 +107,35 @@ public boolean addOrNotify(Translog.Location location, Consumer<Boolean> listene
         return true;
     }
 
+    @Override
+    public void close() throws IOException {
+        List<Tuple<Translog.Location, Consumer<Boolean>>> oldListeners;
+        synchronized (this) {
+            oldListeners = this.refreshListeners;
+            // Setting refreshListeners to emptyList() "closes" it
```

Comment:
I'd rather we used a boolean flag to indicate closed instead of a special reference.